File: cli/src/githubToken.js
// GitHub PAT validation for one Nyx session. Never persist or log token values.

const API_ROOT = 'https://api.github.com'

function targetRepository() {
  return {
    owner: process.env.GITHUB_TARGET_OWNER || 'HelloItsJustin',
    repo: process.env.GITHUB_TARGET_REPO || 'FinForge',
  }
}

function headers(token) {
  return {
    Authorization: `Bearer ${token}`,
    Accept: 'application/vnd.github+json',
    'User-Agent': 'Nyx-CLI/2.0',
    'X-GitHub-Api-Version': '2022-11-28',
  }
}

function scopesFrom(response) {
  return (response.headers.get('x-oauth-scopes') || '').split(',').map(scope => scope.trim()).filter(Boolean)
}

function tokenKind(response) {
  // Header metadata is used only for helpful error text. Fine-grained PATs do
  // not return it, so authorization always uses repository.permissions.push.
  return scopesFrom(response).length ? 'classic' : 'fine-grained'
}

function missingWriteAccess(kind) {
  if (kind === 'classic') {
    return "Token is valid but missing write access — Auto-PR will not work. For a classic token, add the 'repo' scope."
  }
  return 'Token is valid but missing write access — Auto-PR will not work. For a fine-grained token, add this repository and grant Contents: Read and write and Pull requests: Read and write (Metadata: Read-only).'
}

function networkError(error) {
  if (error?.name === 'TimeoutError' || error?.name === 'AbortError') return 'GitHub validation timed out. Check your network connection and try again.'
  return 'GitHub validation could not reach GitHub. Check your network connection and try again.'
}

/**
 * GET /repos/{owner}/{repo} is the sole permission gate. This supports both
 * classic and fine-grained PATs; X-OAuth-Scopes is never an access check.
 */
export async function validateGithubToken(token, target = targetRepository()) {
  const { default: fetch } = await import('node-fetch')
  const request = url => fetch(url, { headers: headers(token), signal: AbortSignal.timeout(15000) })
  const targetName = `${target.owner}/${target.repo}`
  let repositoryResponse
  try {
    repositoryResponse = await request(`${API_ROOT}/repos/${encodeURIComponent(target.owner)}/${encodeURIComponent(target.repo)}`)
  } catch (error) {
    return { ok: false, reason: networkError(error) }
  }
  const kind = tokenKind(repositoryResponse)
  if (repositoryResponse.status === 401) return { ok: false, reason: 'Invalid or expired GitHub token. Paste a current personal access token and try again.' }
  if (repositoryResponse.status === 404) return { ok: false, reason: `Token is valid but cannot read ${targetName}. Add that repository to the token's repository access.` }
  if (repositoryResponse.status === 403) return { ok: false, reason: missingWriteAccess(kind) }
  if (!repositoryResponse.ok) return { ok: false, reason: `GitHub could not validate access to ${targetName} (HTTP ${repositoryResponse.status}).` }

  const repository = await repositoryResponse.json().catch(() => ({}))
  if (!repository?.permissions?.push) return { ok: false, reason: missingWriteAccess(kind) }

  let profileResponse
  try {
    profileResponse = await request(`${API_ROOT}/user`)
  } catch (error) {
    return { ok: false, reason: networkError(error) }
  }
  if (profileResponse.status === 401) return { ok: false, reason: 'Invalid or expired GitHub token. Paste a current personal access token and try again.' }
  if (!profileResponse.ok) return { ok: false, reason: `GitHub validated repository access but could not read the authenticated user (HTTP ${profileResponse.status}).` }
  const profile = await profileResponse.json().catch(() => ({}))
  const scopes = scopesFrom(repositoryResponse)
  return {
    ok: true,
    user: `@${profile.login}`,
    login: profile.login,
    scopes: kind === 'classic' ? `Classic PAT: ${scopes.join(', ')}` : 'Fine-grained PAT: repository write access verified',
    tokenKind: kind,
    permissions: { pull: Boolean(repository.permissions?.pull), push: true },
    target: targetName,
  }
}

export { targetRepository, missingWriteAccess }

File: backend/github_token.py
"""Session-only personal-access-token validation for GitHub Auto-PR."""
from __future__ import annotations

import os
import re
from dataclasses import dataclass
from typing import Any

import httpx

API_ROOT = "https://api.github.com"


@dataclass
class GitHubTokenValidationError(ValueError):
    message: str
    status_code: int = 400

    def __str__(self) -> str:
        return self.message


def configured_target(repo_url: str | None = None) -> tuple[str, str]:
    if repo_url:
        match = re.search(r"github\.com[/:]([^/]+)/([^/#?\s]+?)(?:\.git)?/?(?:[?#].*)?$", repo_url.strip())
        if not match:
            raise GitHubTokenValidationError("Nyx can validate Auto-PR access only for a GitHub repository URL.")
        return match.group(1), match.group(2)
    return os.environ.get("GITHUB_TARGET_OWNER", "HelloItsJustin"), os.environ.get("GITHUB_TARGET_REPO", "FinForge")


def _headers(token: str) -> dict[str, str]:
    return {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "User-Agent": "Nyx-Local/2.0",
        "X-GitHub-Api-Version": "2022-11-28",
    }


def _token_kind(response: httpx.Response) -> str:
    """Only labels errors; X-OAuth-Scopes is never used for authorization."""
    return "classic" if response.headers.get("x-oauth-scopes", "").strip() else "fine-grained"


def _missing_write_access(token_kind: str) -> GitHubTokenValidationError:
    message = (
        "Token is valid but missing write access — Auto-PR will not work. For a classic token, add the 'repo' scope."
        if token_kind == "classic"
        else "Token is valid but missing write access — Auto-PR will not work. For a fine-grained token, add this repository and grant Contents: Read and write and Pull requests: Read and write (Metadata: Read-only)."
    )
    return GitHubTokenValidationError(message, 403)


async def validate_github_token(token: str, repo_url: str | None = None) -> dict[str, Any]:
    """Authorize against the exact target repo; no raw token is persisted or returned."""
    token = token.strip()
    if not token:
        raise GitHubTokenValidationError("A GitHub personal access token is required.")
    owner, repo = configured_target(repo_url)
    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            # This endpoint's permissions.push property works for both PAT kinds.
            repository = await client.get(f"{API_ROOT}/repos/{owner}/{repo}", headers=_headers(token))
            token_kind = _token_kind(repository)
            if repository.status_code == 401:
                raise GitHubTokenValidationError("Invalid or expired GitHub token. Paste a current personal access token and try again.", 401)
            if repository.status_code == 404:
                raise GitHubTokenValidationError(f"Token is valid but cannot read {owner}/{repo}. Add that repository to the token's repository access.", 404)
            if repository.status_code == 403:
                raise _missing_write_access(token_kind)
            if not repository.is_success:
                raise GitHubTokenValidationError(f"GitHub could not validate access to {owner}/{repo} (HTTP {repository.status_code}).", 502)
            permissions = repository.json().get("permissions") or {}
            if not permissions.get("push"):
                raise _missing_write_access(token_kind)

            # Identity is shown only after repo authorization succeeds.
            profile = await client.get(f"{API_ROOT}/user", headers=_headers(token))
            if profile.status_code == 401:
                raise GitHubTokenValidationError("Invalid or expired GitHub token. Paste a current personal access token and try again.", 401)
            if not profile.is_success:
                raise GitHubTokenValidationError(f"GitHub validated repository access but could not read the authenticated user (HTTP {profile.status_code}).", 502)
            login = str(profile.json().get("login") or "")
            if not login:
                raise GitHubTokenValidationError("GitHub returned no authenticated username for this token.", 502)
            scopes = [scope.strip() for scope in repository.headers.get("x-oauth-scopes", "").split(",") if scope.strip()]
            label = f"Classic PAT: {', '.join(scopes)}" if token_kind == "classic" else "Fine-grained PAT: repository write access verified"
            return {
                "login": login,
                "scopes": label,
                "token_kind": token_kind,
                "target": f"{owner}/{repo}",
                "permissions": {"pull": bool(permissions.get("pull")), "push": True},
            }
    except GitHubTokenValidationError:
        raise
    except httpx.TimeoutException as error:
        raise GitHubTokenValidationError("GitHub validation timed out. Check your network connection and try again.", 504) from error
    except httpx.RequestError as error:
        raise GitHubTokenValidationError("GitHub validation could not reach GitHub. Check your network connection and try again.", 503) from error
